plugins/Compressor/Compressor.cpp

```diff
@@ -0,0 +1,648 @@
+		s[1] *= gainResult[1] * m_inGainVal * m_outGainVal * (outBalance < 0 ? 1 + outBalance : 1);
+
+		if (midside)// Convert mid/side back to left/right
+		{
+			const float temp1 = s[0];
+			const float temp2 = s[1] * 0.5;
+			s[0] = temp1 + temp2;
+			s[1] = temp1 - temp2;
+		}
+
+		m_prevOut[0] = s[0];
+		m_prevOut[1] = s[1];
+
+		// Negate wet signal from dry signal
+		if (audition)
+		{
+			s[0] = (-s[0] + delayedDrySignal[0] * m_outGainVal);
+			s[1] = (-s[1] + delayedDrySignal[1] * m_outGainVal);
+		}
+		else if (autoMakeup)
+		{
+			s[0] *= m_autoMakeupVal;
+			s[1] *= m_autoMakeupVal;
+		}
+
+		// Calculate wet/dry value results
+		const float temp1 = delayedDrySignal[0];
+		const float temp2 = delayedDrySignal[1];
+		buf[f][0] = d * temp1 + w * s[0];
+		buf[f][1] = d * temp2 + w * s[1];
+		buf[f][0] = (1 - m_mixVal) * temp1 + m_mixVal * buf[f][0];
+		buf[f][1] = (1 - m_mixVal) * temp2 + m_mixVal * buf[f][1];
+
+		outSum += s[0] + s[1];
+
+		lInPeak = drySignal[0] > lInPeak ? drySignal[0] : lInPeak;
+		rInPeak = drySignal[1] > rInPeak ? drySignal[1] : rInPeak;
+		lOutPeak = s[0] > lOutPeak ? s[0] : lOutPeak;
+		rOutPeak = s[1] > rOutPeak ? s[1] : rOutPeak;
+	}
+
+	checkGate(outSum / frames);
+	m_compressorControls.m_outPeakL = lOutPeak;
+	m_compressorControls.m_outPeakR = rOutPeak;
+	m_compressorControls.m_inPeakL = lInPeak;
+	m_compressorControls.m_inPeakR = rInPeak;
+
+	return isRunning();
+}
+
+
+// Regular modulo doesn't handle negative numbers correctly.  This does.
+inline int CompressorEffect::realmod(int k, int n)
+{
+	return (k %= n) < 0 ? k+n : k;
+}
+
+// Regular fmod doesn't handle negative numbers correctly.  This does.
+inline float CompressorEffect::realfmod(float k, float n)
+{
+	return (k = fmod(k, n)) < 0 ? k+n : k;
+}
+
+
+
+inline void CompressorEffect::calcTiltFilter(sample_t inputSample, sample_t &outputSample, int filtNum)
+{
+	m_tiltOut[filtNum] = m_a0 * inputSample + m_b1 * m_tiltOut[filtNum];
+	outputSample = inputSample + m_lgain * m_tiltOut[filtNum] + m_hgain * (inputSample - m_tiltOut[filtNum]);
+}
+
+
+
+void CompressorEffect::changeSampleRate()
+{
+	m_sampleRate = Engine::mixer()->processingSampleRate();
+
+	m_coeffPrecalc = COMP_LOG / (m_sampleRate * 0.001f);
+
+	// 200 ms
+	m_crestTimeConst = exp(-1.f / (0.2f * m_sampleRate));
+
+	// 20 ms
+	m_lookaheadDelayLength = 0.02 * m_sampleRate;
+	m_inputBuf[0].resize(m_lookaheadDelayLength);
+	m_inputBuf[1].resize(m_lookaheadDelayLength);
+
+	m_lookaheadBuf[0].resize(m_lookaheadDelayLength);
+	m_lookaheadBuf[1].resize(m_lookaheadDelayLength);
+
+	m_preLookaheadBuf[0].resize(m_lookaheadDelayLength);
+	m_preLookaheadBuf[1].resize(m_lookaheadDelayLength);
+
+	calcAutoMakeup();
+	calcAttack();
+	calcRelease();
+	calcRatio();
+	calcRange();
+	calcLookaheadLength();
+	calcHold();
+	resizeRMS();
+	calcThreshold();
+	calcKnee();
+	calcOutGain();
+	calcInGain();
+	calcTiltCoeffs();
+	calcMix();
```

Comment:
Is this order *THE ORDER* it should be done in? If so then a comment saying this order shouldn't change should be added.

Reply:
I didn't think the order mattered at first, but I just noticed there are some incredibly rare niche edgecases in which calcAutoMakeup being before Threshold/Ratio/Knee could potentially cause an issue.  I'll fix that.